### PR TITLE
[hotfix][hold] alert users of pending registration changes

### PR DESCRIPTION
**Pull request by HarryRybacki**
## Purpose
With registration retractions and embargoes going into production on the 29th of this month, users need notification regarding the pending behavior changes of registrations.

## Changes
Users that contribute to registrations will be emailed the template (needs revision @bnosek @saradbowman) regarding the future changes.

Users that wish to create a new registration will be warned of upcoming changes:
screen shot 2015-05-18 at 8 44 01 am

Users that wish to make existing registrations public will also be warned of upcoming changes:
screen shot 2015-05-18 at 8 45 09 am

All contributors to registrations will receive an email akin to what Mr. Pew would see below:
screen shot 2015-05-18 at 9 31 11 am

## Side Effects
None

## Notes
The action labels will need to revised or removed once retractions and embargoes are in production.

@sloria @JeffSpies -- Retractions/Embargoes are set to get merged into production on the 29th. This hotfix and respective email should be merged and issued out as close to tomorrow morning as possible. Changes are as small as I could make them to help facilitate Steve's review.

### CODE REVIEW

**Comment by HarryRybacki** [comment]
@bnosek @saradbowman  @JeffSpies @sloria all messages have been reviewed and revised. Please give them another once over and let me know if anything strikes you as needing further revisions.

**Comment by HarryRybacki** [comment]
@bnosek @saradbowman example email output added to PR description.

**Comment by saradbowman** [comment]
Looks good to me!

File: scripts/email_registration_contributors.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+from website import models
+from website.app import init_app
+from website.mails import Mail, send_mail
+
+
+def main():
+    registrations = models.Node.find(Q('is_registration', 'eq', True))
+    contributors = []
+    for node in registrations:
+        contributors.extend([contrib for contrib in node.contributors if contrib not in contributors])
```
Comment (sloria) [inline scripts/email_registration_contributors.py]: We need to make sure that users only receive this message once. Use the `security_messages` field on a User for this.

Comment (HarryRybacki) [inline scripts/email_registration_contributors.py]: @sloria done

File: website/templates/emails/email_contributors_of_registrations.txt.mako [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+This change removes the possibility of having a permanently private registration.  The change also adds features to promote persistence and preservation.  OSF registrations will receive a persistent identifier (DOI), and registered projects will be archived by the OSF back-end to preserve the content.  Finally, when one project administrator registers the project, all contributors are notified of the registration and all project administrators will have 48 hours to cancel the registration before it is permanent.  We believe that these changes better aligns the registration function with the expectations of what registration should do.
```
Comment (sloria) [inline website/templates/emails/email_contributors_of_registrations.txt.mako]: "These changes better aligns" -> "These changes better align"

I can make this change when I merge.

File: scripts/email_registration_contributors.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+MAILER = Mail(
+    'email_contributors_of_registrations',
+    'Important Update to Registrations on OSF'
+)
+
+
+def send_retraction_and_embargo_addition_message(contrib, label, mail, dry_run=True):
+    if label in contrib.security_messages:
+        return
+    send_mail(contrib.username, mail, user=contrib)
```
Comment (sloria) [inline scripts/email_registration_contributors.py]: We shouldn't send the mail in dry_run mode.

Comment (HarryRybacki) [inline scripts/email_registration_contributors.py]: done

File: website/templates/emails/email_contributors_of_registrations.txt.mako (hunk 2 above)
Comment (HarryRybacki) [inline website/templates/emails/email_contributors_of_registrations.txt.mako]: done

File: scripts/email_registration_contributors.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    for contrib in contributors:
+        if dry_run:
+            logger.info('Dry run mode')
+        logger.info('Sending message to user {0!r}'.format(contrib))
+        send_retraction_and_embargo_addition_message(contrib, MESSAGE_NAME, MAILER, dry_run=dry_run)
+
+
+if __name__ == '__main__':
+    import sys
+    script_utils.add_file_logger(logger, __file__)
```
Comment (sloria) [inline scripts/email_registration_contributors.py]: This should only be added if not in `dry` mode.

File: scripts/email_registration_contributors.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def get_registration_contributors():
+    """Returns set of users that contribute to registrations."""
+    registrations = models.Node.find(Q('is_registration', 'eq', True))
+    contributors = []
+
+    def has_received_message(contrib):
+        query = (Q('_id', 'eq', contrib._id) & Q('security_messages.{0}'.format(MESSAGE_NAME), 'exists', False))
+        return False if len(models.User.find(query)) > 0 else True
```
Comment (sloria) [inline scripts/email_registration_contributors.py]: `if-else` is unnecessary.

```python
return bool(models.User.find(query).count())
```

File: scripts/email_registration_contributors.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+    for node in registrations:
+        contributors.extend([contrib for contrib in node.contributors if contrib not in contributors and not has_received_message(contrib)])
+
```
Comment (sloria) [inline scripts/email_registration_contributors.py]: Might be good to log how many contributors will receive the email.

File: scripts/email_registration_contributors.py (hunk 5 above)
Comment (HarryRybacki) [inline scripts/email_registration_contributors.py]: it's actually the compliment of that -- I think `return models.User.find(query).count() == 0` is more concise yet still clear. `return not bool(models.User.find(query).count())` just looked weird

File: scripts/email_registration_contributors.py (hunk 6 above)
Comment (HarryRybacki) [inline scripts/email_registration_contributors.py]: done

File: website/templates/emails/email_contributors_of_registrations.txt.mako [hunk 7]
```diff
@@ -0,0 +1,18 @@
+Dear ${user.fullname}, 
+
+You are receiving this message because you have an account on the Open Science Framework (OSF).  We are writing to let you know of a change to our model for registrations that may affect how you choose to use that feature.
```
Comment (HarryRybacki) [inline website/templates/emails/email_contributors_of_registrations.txt.mako]: @bnosek @sloria Need to update this with however we want to work the tentative dates

**Comment by HarryRybacki** [comment]
@sloria good to go pending a language review from @bnosek. How do we want to handle the action message updates?

**Pull request merged**